filters/encrypt/filter.go

```diff
@@ -0,0 +1,542 @@
+	const op = "event.(Filter).Process"
+	if e == nil {
+		return nil, fmt.Errorf("%s: missing event: %w", op, ErrInvalidParameter)
+	}
+	if i, ok := e.Payload.(RotateWrapper); ok {
+		ef.l.Lock()
+		defer ef.l.Unlock()
+		if i.Wrapper() != nil {
+			ef.Wrapper = i.Wrapper()
+		}
+		if i.HmacSalt() != nil {
+			ef.HmacSalt = make([]byte, len(i.HmacSalt()))
+			copy(ef.HmacSalt, i.HmacSalt())
+		}
+		if i.HmacInfo() != nil {
+			ef.HmacInfo = make([]byte, len(i.HmacInfo()))
+			copy(ef.HmacInfo, i.HmacInfo())
+		}
+		return nil, nil
+	}
+
+	var opts []Option
+	var optWrapper wrapping.Wrapper
+	if i, ok := e.Payload.(EventWrapperInfo); ok {
+		ef.l.RLock()
+		w, err := NewEventWrapper(ef.Wrapper, i.EventId())
+		ef.l.RUnlock()
+		if err != nil {
+			return nil, fmt.Errorf("%s: %w", op, err)
+		}
+		optWrapper := w
+		opts = append(opts, WithWrapper(optWrapper))
+		opts = append(opts, WithInfo(i.HmacInfo()))
+		opts = append(opts, WithSalt(i.HmacSalt()))
+	}
+
+	if ef.Wrapper == nil && optWrapper == nil {
+		return nil, fmt.Errorf("%s: missing wrapper: %w", op, ErrInvalidParameter)
+	}
+
+	if e.Payload == nil {
+		return e, nil
+	}
+
+	// Get both the value and the type of what the payload points to. Value is
+	// used to mutate underlying data and Type is used to get the name of the
+	// field.
+	payloadValue := reflect.ValueOf(e.Payload)
+	switch payloadValue.Kind() {
+	case reflect.Ptr, reflect.Interface:
+		payloadValue = reflect.ValueOf(e.Payload).Elem()
+	}
+
+	pType := payloadValue.Type()
+	pKind := payloadValue.Kind()
+
+	taggedInterface, isTaggable := payloadValue.Interface().(Taggable)
+
+	switch {
+	case pType == reflect.TypeOf("") || pType == reflect.TypeOf([]uint8{}):
+		if !payloadValue.CanSet() {
+			return nil, fmt.Errorf("%s: unable to redact string payload (not setable): %w", op, ErrInvalidParameter)
+		}
+		ef.l.RLock()
+		classificationTag := getClassificationFromTagString(string(SecretClassification), withFilterOperations(ef.FilterOperationOverrides))
+		ef.l.RUnlock()
+		if err := ef.filterValue(ctx, payloadValue, classificationTag, opts...); err != nil {
+			return nil, fmt.Errorf("%s: %w", op, err)
+		}
+	case isTaggable:
+		if err := ef.filterTaggable(ctx, taggedInterface, opts...); err != nil {
+			return nil, fmt.Errorf("%s: %w", op, err)
+		}
+	case pKind == reflect.Slice:
+		switch {
+		// if the field is a slice of string or slice of []byte
+		case pType == reflect.TypeOf([]string{}) || pType == reflect.TypeOf([]*string{}) || pType == reflect.TypeOf([][]uint8{}):
+			ef.l.RLock()
+			classificationTag := getClassificationFromTagString(string(SecretClassification), withFilterOperations(ef.FilterOperationOverrides))
+			ef.l.RUnlock()
+			if err := ef.filterSlice(ctx, classificationTag, payloadValue, opts...); err != nil {
+				return nil, fmt.Errorf("%s: %w", op, err)
+			}
+		// if the field is a slice of structs, recurse through them...
```

Comment:
This is iteration, not recursion. I realize this is a nit but I forgot that this was the `Process` function and got confused when the next function was `filterField`. :-)

Reply:
I guess your point is that `filterField` will recurse, but that actually makes me wonder about the difference between this logic in Process and the logic in `filterField`. Should the top level value simply be sent to `filterField` from the get-go?

Reply:
I did an implementation like that and it was far more complex, so I went with this sort of "mixed" version that's more explicit and only recurses struct fields